feat: add concept ex for smart pointers (#846)
exercism/cpp: exercises/concept/power-of-troy/.docs/instructions.md

# Instructions

This exercise takes you to the world of Troy.
The lives of its people are full of wonder and magic.
Many humans in Troy possess _powers_, that are used frequently in their daily lives.
Powers are used to re-shape the world or influence Troy's fauna and other people.
Magic also manifests in _unique artifacts_, that are highly sought after by adventurers, artisans and sages.

In this exercise you are going to write code to model the humans of Troy, their possessed artifacts and power interactions.

You have six tasks.
The first one is related to creating a human, the other five are about handling powers and artifacts.

## 1. Bring humans to the world of Troy

For your model of Troy humans are the most important feature.
Your model human should be able to possess a _unique artifact_.
They should also have the ability to manifest a _power_.
These powers might affect other humans, so you also want to model if a human is influenced by some other power.

You are provided with basic implementations of `artifact` and `power` structs.

> Implement a `human` struct (or class) that has a _smart-pointer_ to an `artifact` as `possession` member variable.

Reviewer: ```suggestion
Implement a `human` struct (or class) that has a _smart-pointer_ to an `artifact` as a `possession` member variable.
```

Each artifact can only be possessed by a single human at any given time.

The `human` should also have variables for their `own_power` and `influenced_by`, which should be _smart-pointers_ to `powers`.
Each `power` might be owned by a single human, but also influence other humans at the same time.

> By default, humans are born without any artifact and neither own any powers nor are they influenced by them.

Reviewer: ```suggestion
By default, humans are born without any artifacts, have no powers and are not under the influence of other powers.
```

```cpp
human mindy_mccready{};
mindy_mccready.possession;
// => nullptr
mindy_mccready.own_power;
// => nullptr
mindy_mccready.influenced_by;
// => nullptr
```

## 2. Bring Artifacts into Troy

Your model is boring without the interaction of its parts.
You want to create unique artifacts and give them to certain humans.

Define the function `give_new_artifact` which returns nothing but takes a `human` and a `string`.

> With the `string` it should define a new `artifact` object and set the `possession` pointer of the `human` accordingly.

Reviewer: ```suggestion
It should define a new `artifact` object using it name, and set the `possession` pointer of the `human` accordingly.
```

The function should not return anything.

```cpp
human erik_magnus_lehnsherr{};
give_new_artifact(erik_magnus_lehnsherr, "Mind shielding helmet");

erik_magnus_lehnsherr.possession->name;
// "Mind shielding helmet"
```

> ## 3. Make items tradeable

Reviewer: It's probably fine, but GH loses syntax highlighting at this point

The world of Troy is all about interaction.
You want people to make trades by exchanging their possessions.

> Write a function `exchange_artifacts` that returns nothing but takes two artifact smart-pointers to exchange the items.

Reviewer: ```suggestion
Write a function `exchange_artifacts` that takes two artifact smart-pointers to exchange the items. 
It does not return anything.
```
